snap lxd daemon - SIGSEGV, invalid memory address or nil pointer dereference

LXD Daemon in snap wont start, keeps looping "database locked"

 * Distribution: Ubuntu 
 * Distribution version: 18.04.2 LTS
 * The output of "lxc info" or if that fails:
   * Kernel version: Linux hetz-hel1 4.15.0-55-generic #60-Ubuntu SMP Tue Jul 2 18:22:20 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
   * LXC version: snap 3.15
   * LXD version: snap 3.15
   * Storage backend in use: zfs

# Issue description

LXD socket unreachable from LXC as lxd daemon fails to start (SIGSEGV)

```
-- Logs begin at Mon 2018-08-27 01:22:02 CAT, end at Wed 2019-07-31 12:45:11 CAT. --
Jul 31 12:38:11 hetz-hel1 lxd.daemon[9273]: => Preparing the system
Jul 31 12:38:11 hetz-hel1 lxd.daemon[9273]: ==> Loading snap configuration
Jul 31 12:38:11 hetz-hel1 lxd.daemon[9273]: ==> Setting up mntns symlink (mnt:[4026532239])
Jul 31 12:38:15 hetz-hel1 lxd.daemon[9273]: ==> Setting up kmod wrapper
Jul 31 12:38:15 hetz-hel1 lxd.daemon[9273]: ==> Preparing /boot
Jul 31 12:38:19 hetz-hel1 lxd.daemon[9273]: ==> Preparing a clean copy of /run
Jul 31 12:38:23 hetz-hel1 lxd.daemon[9273]: ==> Preparing a clean copy of /etc
Jul 31 12:38:43 hetz-hel1 lxd.daemon[9273]: ==> Setting up ceph configuration
Jul 31 12:38:43 hetz-hel1 lxd.daemon[9273]: ==> Setting up LVM configuration
Jul 31 12:38:43 hetz-hel1 lxd.daemon[9273]: ==> Rotating logs
Jul 31 12:38:43 hetz-hel1 lxd.daemon[9273]: ==> Setting up ZFS (0.7)
Jul 31 12:38:43 hetz-hel1 lxd.daemon[9273]: ==> Escaping the systemd cgroups
Jul 31 12:38:43 hetz-hel1 lxd.daemon[9273]: ==> Escaping the systemd process resource limits
Jul 31 12:38:43 hetz-hel1 lxd.daemon[9273]: ==> Enabling CRIU
Jul 31 12:38:43 hetz-hel1 lxd.daemon[9273]: ==> Disabling shiftfs on this kernel (auto)
Jul 31 12:38:47 hetz-hel1 lxd.daemon[9273]: => Starting LXCFS
Jul 31 12:38:47 hetz-hel1 lxd.daemon[9273]: => Starting LXD
Jul 31 12:38:51 hetz-hel1 lxd.daemon[9273]: mount namespace: 5
Jul 31 12:38:51 hetz-hel1 lxd.daemon[9273]: hierarchies:
Jul 31 12:38:51 hetz-hel1 lxd.daemon[9273]:   0: fd:   6: memory
Jul 31 12:38:51 hetz-hel1 lxd.daemon[9273]:   1: fd:   7: perf_event
Jul 31 12:38:51 hetz-hel1 lxd.daemon[9273]:   2: fd:   8: cpu,cpuacct
Jul 31 12:38:51 hetz-hel1 lxd.daemon[9273]:   3: fd:   9: cpuset
Jul 31 12:38:51 hetz-hel1 lxd.daemon[9273]:   4: fd:  10: rdma
Jul 31 12:38:51 hetz-hel1 lxd.daemon[9273]:   5: fd:  11: blkio
Jul 31 12:38:51 hetz-hel1 lxd.daemon[9273]:   6: fd:  12: devices
Jul 31 12:38:51 hetz-hel1 lxd.daemon[9273]:   7: fd:  13: hugetlb
Jul 31 12:38:51 hetz-hel1 lxd.daemon[9273]:   8: fd:  14: freezer
Jul 31 12:38:51 hetz-hel1 lxd.daemon[9273]:   9: fd:  15: pids
Jul 31 12:38:51 hetz-hel1 lxd.daemon[9273]:  10: fd:  16: net_cls,net_prio
Jul 31 12:38:51 hetz-hel1 lxd.daemon[9273]:  11: fd:  17: name=systemd
Jul 31 12:38:51 hetz-hel1 lxd.daemon[9273]:  12: fd:  18: unified
Jul 31 12:38:56 hetz-hel1 lxd.daemon[9273]: t=2019-07-31T12:38:56+0200 lvl=eror msg="Failed to start image leftover cleanup operation" err="failed to add operation f48e3a40-4328-4b93-8768-d9757c55db98 to database: database is locked"
Jul 31 12:38:56 hetz-hel1 lxd.daemon[9273]: panic: runtime error: invalid memory address or nil pointer dereference
Jul 31 12:38:56 hetz-hel1 lxd.daemon[9273]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x58 pc=0xe760ca]
Jul 31 12:38:56 hetz-hel1 lxd.daemon[9273]: goroutine 1 [running]:
Jul 31 12:38:56 hetz-hel1 lxd.daemon[9273]: main.(*operation).Run(0x0, 0x1c, 0x0, 0x0)
Jul 31 12:38:56 hetz-hel1 lxd.daemon[9273]:         /build/lxd/parts/lxd/go/src/github.com/lxc/lxd/lxd/operations.go:134 +0x3a
Jul 31 12:38:56 hetz-hel1 lxd.daemon[9273]: main.pruneLeftoverImages(0xc0001fefc0)
Jul 31 12:38:56 hetz-hel1 lxd.daemon[9273]:         /build/lxd/parts/lxd/go/src/github.com/lxc/lxd/lxd/images.go:1215 +0xfb
Jul 31 12:38:56 hetz-hel1 lxd.daemon[9273]: main.(*Daemon).init(0xc0001fefc0, 0x4107e0, 0x60)
Jul 31 12:38:56 hetz-hel1 lxd.daemon[9273]:         /build/lxd/parts/lxd/go/src/github.com/lxc/lxd/lxd/daemon.go:794 +0x12c8
Jul 31 12:38:56 hetz-hel1 lxd.daemon[9273]: main.(*Daemon).Init(0xc0001fefc0, 0xc0001fefc0, 0xc00022c780)
Jul 31 12:38:56 hetz-hel1 lxd.daemon[9273]:         /build/lxd/parts/lxd/go/src/github.com/lxc/lxd/lxd/daemon.go:496 +0x2f
Jul 31 12:38:56 hetz-hel1 lxd.daemon[9273]: main.(*cmdDaemon).Run(0xc000171da0, 0xc000287180, 0xc000175400, 0x0, 0x4, 0x0, 0x0)
Jul 31 12:38:56 hetz-hel1 lxd.daemon[9273]:         /build/lxd/parts/lxd/go/src/github.com/lxc/lxd/lxd/main_daemon.go:65 +0x417
Jul 31 12:38:56 hetz-hel1 lxd.daemon[9273]: github.com/spf13/cobra.(*Command).execute(0xc000287180, 0xc000032060, 0x4, 0x4, 0xc000287180, 0xc000032060)
Jul 31 12:38:56 hetz-hel1 lxd.daemon[9273]:         /build/lxd/parts/lxd/go/src/github.com/spf13/cobra/command.go:826 +0x465
Jul 31 12:38:56 hetz-hel1 lxd.daemon[9273]: github.com/spf13/cobra.(*Command).ExecuteC(0xc000287180, 0xc00027df60, 0x1, 0x1)
Jul 31 12:38:56 hetz-hel1 lxd.daemon[9273]:         /build/lxd/parts/lxd/go/src/github.com/spf13/cobra/command.go:914 +0x2fc
Jul 31 12:38:56 hetz-hel1 lxd.daemon[9273]: github.com/spf13/cobra.(*Command).Execute(...)
Jul 31 12:38:56 hetz-hel1 lxd.daemon[9273]:         /build/lxd/parts/lxd/go/src/github.com/spf13/cobra/command.go:864
Jul 31 12:38:56 hetz-hel1 lxd.daemon[9273]: main.main()
Jul 31 12:38:56 hetz-hel1 lxd.daemon[9273]:         /build/lxd/parts/lxd/go/src/github.com/lxc/lxd/lxd/main.go:176 +0xe82
Jul 31 12:38:56 hetz-hel1 lxd.daemon[9273]: => LXD failed to start
Jul 31 12:40:22 hetz-hel1 lxd.daemon[9273]: mount namespace: 7
Jul 31 12:40:22 hetz-hel1 lxd.daemon[9273]: hierarchies:
Jul 31 12:40:22 hetz-hel1 lxd.daemon[9273]:   0: fd:   8: memory
Jul 31 12:40:22 hetz-hel1 lxd.daemon[9273]:   1: fd:   9: perf_event
Jul 31 12:40:22 hetz-hel1 lxd.daemon[9273]:   2: fd:  10: cpu,cpuacct
Jul 31 12:40:22 hetz-hel1 lxd.daemon[9273]:   3: fd:  11: cpuset
Jul 31 12:40:22 hetz-hel1 lxd.daemon[9273]:   4: fd:  12: rdma
Jul 31 12:40:22 hetz-hel1 lxd.daemon[9273]:   5: fd:  13: blkio
Jul 31 12:40:22 hetz-hel1 lxd.daemon[9273]:   6: fd:  14: devices
Jul 31 12:40:22 hetz-hel1 lxd.daemon[9273]:   7: fd:  15: hugetlb
Jul 31 12:40:22 hetz-hel1 lxd.daemon[9273]:   8: fd:  16: freezer
Jul 31 12:40:22 hetz-hel1 lxd.daemon[9273]:   9: fd:  17: pids
Jul 31 12:40:22 hetz-hel1 lxd.daemon[9273]:  10: fd:  18: net_cls,net_prio
Jul 31 12:40:22 hetz-hel1 lxd.daemon[9273]:  11: fd:  19: name=systemd
Jul 31 12:40:22 hetz-hel1 lxd.daemon[9273]:  12: fd:  20: unified
Jul 31 12:40:22 hetz-hel1 lxd.daemon[9273]: lxcfs.c: 152: do_reload: lxcfs: reloaded
```
## Flood of dmesg spam (repeated several times over as lxd attempts to restart) not sure if relevant
```
[Wed Jul 31 12:40:17 2019] new mount options do not match the existing superblock, will be ignored
```
`IP_ADDR` = Public IP removed.
### /var/snap/lxd/common/lxd/logs/lxd.log
```
t=2019-07-31T12:40:23+0200 lvl=info msg="LXD 3.15 is starting in normal mode" path=/var/snap/lxd/common/lxd
t=2019-07-31T12:40:23+0200 lvl=info msg="Kernel uid/gid map:" 
t=2019-07-31T12:40:23+0200 lvl=info msg=" - u 0 0 4294967295" 
t=2019-07-31T12:40:23+0200 lvl=info msg=" - g 0 0 4294967295" 
t=2019-07-31T12:40:23+0200 lvl=info msg="Configured LXD uid/gid map:" 
t=2019-07-31T12:40:23+0200 lvl=info msg=" - u 0 1000000 1000000000" 
t=2019-07-31T12:40:23+0200 lvl=info msg=" - g 0 1000000 1000000000" 
t=2019-07-31T12:40:23+0200 lvl=info msg="Kernel features:" 
t=2019-07-31T12:40:23+0200 lvl=info msg=" - netnsid-based network retrieval: no" 
t=2019-07-31T12:40:23+0200 lvl=info msg=" - uevent injection: no" 
t=2019-07-31T12:40:23+0200 lvl=info msg=" - seccomp listener: no" 
t=2019-07-31T12:40:23+0200 lvl=info msg=" - unprivileged file capabilities: yes" 
t=2019-07-31T12:40:23+0200 lvl=info msg=" - shiftfs support: no" 
t=2019-07-31T12:40:23+0200 lvl=info msg="Initializing local database" 
t=2019-07-31T12:40:23+0200 lvl=info msg="Starting /dev/lxd handler:" 
t=2019-07-31T12:40:23+0200 lvl=info msg=" - binding devlxd socket" socket=/var/snap/lxd/common/lxd/devlxd/sock
t=2019-07-31T12:40:23+0200 lvl=info msg="REST API daemon:" 
t=2019-07-31T12:40:23+0200 lvl=info msg=" - binding Unix socket" socket=/var/snap/lxd/common/lxd/unix.socket
t=2019-07-31T12:40:23+0200 lvl=info msg=" - binding TCP socket" socket=IP_ADDR:8443
t=2019-07-31T12:40:23+0200 lvl=info msg="Initializing global database" 
t=2019-07-31T12:40:23+0200 lvl=info msg="Initializing storage pools" 
t=2019-07-31T12:40:23+0200 lvl=info msg="Initializing networks" 
t=2019-07-31T12:40:29+0200 lvl=eror msg="Failed to start image leftover cleanup operation" err="failed to add operation 1aee3c3b-658b-4ea4-bd31-6183108190ee to database: database is locked"
t=2019-07-31T12:40:29+0200 lvl=info msg="Pruning leftover image files" 
```

`lxd --debug --group lxd 2>&1 | tee lxd_debug_fail.log`
```
t=2019-07-31T12:48:31+0200 lvl=dbug msg="Connecting to a local LXD over a Unix socket" 
t=2019-07-31T12:48:31+0200 lvl=dbug msg="Sending request to LXD" etag= method=GET url=http://unix.socket/1.0
t=2019-07-31T12:48:31+0200 lvl=info msg="LXD 3.15 is starting in normal mode" path=/var/snap/lxd/common/lxd
t=2019-07-31T12:48:31+0200 lvl=info msg="Kernel uid/gid map:" 
t=2019-07-31T12:48:31+0200 lvl=info msg=" - u 0 0 4294967295" 
t=2019-07-31T12:48:31+0200 lvl=info msg=" - g 0 0 4294967295" 
t=2019-07-31T12:48:31+0200 lvl=info msg="Configured LXD uid/gid map:" 
t=2019-07-31T12:48:31+0200 lvl=info msg=" - u 0 1000000 1000000000" 
t=2019-07-31T12:48:31+0200 lvl=info msg=" - g 0 1000000 1000000000" 
t=2019-07-31T12:48:31+0200 lvl=info msg="Kernel features:" 
t=2019-07-31T12:48:31+0200 lvl=info msg=" - netnsid-based network retrieval: no" 
t=2019-07-31T12:48:31+0200 lvl=info msg=" - uevent injection: no" 
t=2019-07-31T12:48:31+0200 lvl=info msg=" - seccomp listener: no" 
t=2019-07-31T12:48:31+0200 lvl=info msg=" - unprivileged file capabilities: yes" 
t=2019-07-31T12:48:31+0200 lvl=info msg=" - shiftfs support: no" 
t=2019-07-31T12:48:31+0200 lvl=info msg="Initializing local database" 
t=2019-07-31T12:48:31+0200 lvl=dbug msg="Initializing database gateway" 
t=2019-07-31T12:48:31+0200 lvl=dbug msg="Start database node" address=IP_ADDR:8443 id=1
t=2019-07-31T12:48:31+0200 lvl=dbug msg="Connecting to a local LXD over a Unix socket" 
t=2019-07-31T12:48:31+0200 lvl=dbug msg="Sending request to LXD" etag= method=GET url=http://unix.socket/1.0
t=2019-07-31T12:48:31+0200 lvl=dbug msg="Detected stale unix socket, deleting" 
t=2019-07-31T12:48:31+0200 lvl=dbug msg="Detected stale unix socket, deleting" 
t=2019-07-31T12:48:31+0200 lvl=info msg="Starting /dev/lxd handler:" 
t=2019-07-31T12:48:31+0200 lvl=info msg=" - binding devlxd socket" socket=/var/snap/lxd/common/lxd/devlxd/sock
t=2019-07-31T12:48:31+0200 lvl=info msg="REST API daemon:" 
t=2019-07-31T12:48:31+0200 lvl=info msg=" - binding Unix socket" socket=/var/snap/lxd/common/lxd/unix.socket
t=2019-07-31T12:48:31+0200 lvl=info msg=" - binding TCP socket" socket=IP_ADDR:8443
t=2019-07-31T12:48:31+0200 lvl=info msg="Initializing global database" 
t=2019-07-31T12:48:31+0200 lvl=dbug msg="Found cert" name=0
t=2019-07-31T12:48:31+0200 lvl=dbug msg="Found cert" name=0
t=2019-07-31T12:48:31+0200 lvl=dbug msg="Dqlite: connected address=IP_ADDR:8443 attempt=0" 
t=2019-07-31T12:48:31+0200 lvl=info msg="Initializing storage pools" 
t=2019-07-31T12:48:31+0200 lvl=dbug msg="Initializing and checking storage pool \"local\"" 
t=2019-07-31T12:48:31+0200 lvl=dbug msg="Checking ZFS storage pool \"local\"" 
t=2019-07-31T12:48:31+0200 lvl=info msg="Initializing networks" 
t=2019-07-31T12:48:36+0200 lvl=dbug msg="Dqlite: closing client" 
t=2019-07-31T12:48:36+0200 lvl=dbug msg="Database error: &errors.errorString{s:\"driver: bad connection\"}" 
t=2019-07-31T12:48:36+0200 lvl=dbug msg="Retry failed db interaction (driver: bad connection)" 
t=2019-07-31T12:48:36+0200 lvl=dbug msg="Found cert" name=0
t=2019-07-31T12:48:36+0200 lvl=dbug msg="Found cert" name=0
t=2019-07-31T12:48:36+0200 lvl=dbug msg="Dqlite: connected address=IP_ADDR:8443 attempt=0" 
t=2019-07-31T12:48:36+0200 lvl=dbug msg="Database error: bindings.Error{Code:5, Message:\"database is locked\"}" 
t=2019-07-31T12:48:36+0200 lvl=dbug msg="Retry failed db interaction (database is locked)" 
t=2019-07-31T12:48:36+0200 lvl=dbug msg="Database error: bindings.Error{Code:5, Message:\"database is locked\"}" 
t=2019-07-31T12:48:36+0200 lvl=dbug msg="Retry failed db interaction (database is locked)" 
t=2019-07-31T12:48:37+0200 lvl=dbug msg="Database error: bindings.Error{Code:5, Message:\"database is locked\"}" 
t=2019-07-31T12:48:37+0200 lvl=dbug msg="Retry failed db interaction (database is locked)" 
t=2019-07-31T12:48:37+0200 lvl=dbug msg="Database error: bindings.Error{Code:5, Message:\"database is locked\"}" 
t=2019-07-31T12:48:37+0200 lvl=dbug msg="Retry failed db interaction (database is locked)" 
t=2019-07-31T12:48:37+0200 lvl=eror msg="Failed to start image leftover cleanup operation" err="failed to add operation 75f53de0-5b50-4d0f-8be0-3a7b597bfc08 to database: database is locked"
t=2019-07-31T12:48:37+0200 lvl=info msg="Pruning leftover image files" 
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x58 pc=0xe760ca]

goroutine 1 [running]:
main.(*operation).Run(0x0, 0x1c, 0x0, 0x0)
        /build/lxd/parts/lxd/go/src/github.com/lxc/lxd/lxd/operations.go:134 +0x3a
main.pruneLeftoverImages(0xc0001c8d80)
        /build/lxd/parts/lxd/go/src/github.com/lxc/lxd/lxd/images.go:1215 +0xfb
main.(*Daemon).init(0xc0001c8d80, 0x4107e0, 0x60)
        /build/lxd/parts/lxd/go/src/github.com/lxc/lxd/lxd/daemon.go:794 +0x12c8
main.(*Daemon).Init(0xc0001c8d80, 0xc0001c8d80, 0xc0001f4780)
        /build/lxd/parts/lxd/go/src/github.com/lxc/lxd/lxd/daemon.go:496 +0x2f
main.(*cmdDaemon).Run(0xc000175c80, 0xc000253180, 0xc0002de6f0, 0x0, 0x3, 0x0, 0x0)
        /build/lxd/parts/lxd/go/src/github.com/lxc/lxd/lxd/main_daemon.go:65 +0x417
github.com/spf13/cobra.(*Command).execute(0xc000253180, 0xc000032110, 0x3, 0x3, 0xc000253180, 0xc000032110)
        /build/lxd/parts/lxd/go/src/github.com/spf13/cobra/command.go:826 +0x465
github.com/spf13/cobra.(*Command).ExecuteC(0xc000253180, 0xc00024bf60, 0x1, 0x1)
        /build/lxd/parts/lxd/go/src/github.com/spf13/cobra/command.go:914 +0x2fc
github.com/spf13/cobra.(*Command).Execute(...)
        /build/lxd/parts/lxd/go/src/github.com/spf13/cobra/command.go:864
main.main()
        /build/lxd/parts/lxd/go/src/github.com/lxc/lxd/lxd/main.go:176 +0xe82
```

--- a/lxd/images.go
+++ b/lxd/images.go
@@ -1209,12 +1209,14 @@ func pruneLeftoverImages(d *Daemon) {
 	op, err := operationCreate(d.cluster, "", operationClassTask, db.OperationImagesPruneLeftover, nil, nil, opRun, nil, nil)
 	if err != nil {
 		logger.Error("Failed to start image leftover cleanup operation", log.Ctx{"err": err})
+		return
 	}
 
 	logger.Infof("Pruning leftover image files")
 	_, err = op.Run()
 	if err != nil {
 		logger.Error("Failed to prune leftover image files", log.Ctx{"err": err})
+		return
 	}
 	logger.Infof("Done pruning leftover image files")
 }

